burn-core/src/nn/rnn/lstm.rs

```diff
@@ -110,7 +124,7 @@ impl<B: Backend> Lstm<B> {
     /// Parameters:
     ///     batched_input: The input tensor of shape [batch_size, sequence_length, input_size].
     ///     state: An optional tuple of tensors representing the initial cell state and hidden state.
-    ///            Each state tensor has shape [batch_size, hidden_size].
+    ///            Each state tensor has shape [num_directions, batch_size, hidden_size].
```

Comment:
I feel like it would be better to have a `BiDirectionalLstm` module instead of one module that does both. The bi-directional module might just have two Lstm modules instead of duplicating all the gates. We could also make it more Rusty by using Kind types:

```rust
pub trait LstmKind<B: Backend> {
    fn forward(...) -> ...;
}
#[derive(Module)]
pub struct UniDirectional{
    input_gate: GateController<B>,
    forget_gate: GateController<B>,
    output_gate: GateController<B>,
    cell_gate: GateController<B>,
};

#[derive(Module)]
pub struct BiDirectional<B: Backend> {
   forward: Lstm<B, UniDirectional>,
   backward: Lstm<B, UniFirectional>,
}
pub struct Lstm<B: Backend, K: LstmKind<B> = UniDirectional<B>> {
   state: K,
}

impl<B: Backend, K: LstmKind<B>> Lstm<B, K> {
   pub fn forward(...) -> ... {
       self.kind.forward(...) // static dispatch to the right forward pass depending on the kind.
   }
}

impl<B: Backend> LstmKind<B> for UniDirectional<B> {
   fn forward(...) -> ... {
       // uni directional forward pass
   }
}
impl<B: Backend> LstmKind<B> for BiDirectional<B> {
   fn forward(...) -> ... {
       // bi directional forward pass
   }
}
```

This is equivalent to having two different modules for the uni directional and bi directional lstm but with syntax sugar so users can also use `Lstm<B>` for uni directional or `Lstm<B, BiDirectional>` when they want both directions. This is the same pattern that we use for the Tensor API.

Reply:
This would also make it backward compatible in term of code, but not in term of state (Recorder).

Reply:
@agelas what are your thoughts

Reply:
@nathanielsimard I agree that this pattern fits better with the overall approach that is usually taken. I think composing LSTM modules like this is a bit more elegant than having various toggles in the `forward` pass, and we can reuse most of the prior implementation that way.

Reply:
@nathanielsimard The problem I have is that the `forward()` for `UniDirectional` and `BiDirectional` have different function signatures due to the inconsistency in the shape of the `state`. This seems not easy to resolve, so it's probably best to keep the LSTM implementation unchanged and let users implement bidirectional LSTM or multi-layer LSTM on their own.

Reply:
Not sure how often bidirectional LSTM is used, but could we have an implementation totally separated from unidirectional LSTM, and just offer both? Because I do agree that messing with function signatures might get a bit bothersome

Reply:
The pattern I'm proposing creates two different types, so there are no problems with the type signature. It's just a way for both types to use the LSTM nomenclature. It's probably easier to start with two different types, then we can add the pattern afterward if it makes sense.

Reply:
@nathanielsimard This pattern is a bit tricky for me, so I think it's best for me to create a new type for bidirectional LSTM. I sincerely hope you can improve it if possible. Thank you very much!